__tests__/authenticationFlow.test.js

```diff
@@ -0,0 +1,324 @@
+      const env = createEnvWithLegacyUser()
+      env.loginByUsernamePassword('flicknote', 'pwd').mustSucceed()
+      return env.verify()
+    })
+    it('can authenticate with email and password', () => {
+      const env = createEnvWithLegacyUser()
+      env.loginByUsernamePassword('flicknote@bemuse.ninja', 'pwd').mustSucceed()
+      return env.verify()
+    })
+    it('cannot authenticate with email if password is wrong', () => {
+      const env = createEnvWithLegacyUser()
+      env.loginByUsernamePassword('flicknote', 'pwdz').mustFail()
+      return env.verify()
+    })
+    it('cannot authenticate with username if password is wrong', () => {
+      const env = createEnvWithLegacyUser()
+      env.loginByUsernamePassword('flicknote@bemuse.ninja', 'pwdz').mustFail()
+      return env.verify()
+    })
+    it('cannot authenticate if unknown player', () => {
+      const env = createEnvWithLegacyUser()
+      env.loginByUsernamePassword('meow', 'pwd').mustFail()
+      return env.verify()
+    })
+    it('should link the player after successful')
+  })
+
+  describe('sign up with username, email and password', () => {
+    it('can sign up as new user', () => {
+      const env = createEnv()
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      env.loginByUsernamePassword('DJTHAI', 'strongpassword').mustSucceed()
+      env.loginByUsernamePassword('thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      return env.verify()
+    })
+    it('cannot sign up if legacy user already exist', () => {
+      const env = createEnv()
+      env.givenLegacyUser('DJTHAI', 'thai@another.place', 'wow')
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustFail()
+      return env.verify()
+    })
+    it('cannot sign up if email already exist as legacy user', () => {
+      const env = createEnv()
+      env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+      env.givenPlayer('THAI')
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustFail()
+      return env.verify()
+    })
+    it('should link the player after successful', () => {
+      const env = createEnv()
+      env.signUp('DJTHAI', 'thai@bemuse.ninja', 'strongpassword').mustSucceed()
+      env.playerWithName('DJTHAI').shouldBeLinkedTo('thai@bemuse.ninja')
+      return env.verify()
+    })
+  })
+
+  describe('edge cases', () => {
+    describe('hijacking user by creating account with existing player id as username', () => {
+      it('should be prevented', () => {
+        const env = createEnv()
+        env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+        env.givenPlayer('THAI')
+        env.externalSignUp('player0', 'player0@bemuse.ninja', 'strongpassword').mustFail()
+        return env.verify()
+      })
+    })
+    describe('an unlinked account (somehow)', () => {
+      it('should try to link with legacy user via email, if possible')
```

Comment:
Nope, an unlinked account can’t be created in the first place if its email clashes with a legacy user’s email.